File: 1337/1337.py
'''
<формула> ::= <число> <последовательность> {"+" <число> <последовательность>}
<последовательность> ::= <часть> <число> {<часть> <число>}
<часть> ::= <химический элемент> | "(" <последовательность> ")"
<химический элемент> ::= <прописная буква> [<строчная буква>]
<прописная буква> ::= "A".."Z"
<строчная буква> ::= "a".."z"
<число> ::= ["1".."9" {"0".."9"}]
'''

from collections import Counter

def readFormula():
    s = input()
    ch = ''
    pos = 0
    EOS = '\n'  # end of string

    def nextChar():
        nonlocal ch, pos

        if pos < len(s):
            ch = s[pos]
            pos += 1
        else:
            ch = EOS

    # <число> ::= ["1".."9" {"0".."9"}]
    def number():
        if len(ch) == 1 and '1' <= ch <= '9':
            num = ord(ch) - ord('0')
            nextChar()

            while '0' <= ch <= '9':
                num = num * 10 + (ord(ch) - ord('0'))
                assert num <= 10_000
                nextChar()

            return num
        else:
            return 1

    # <химический элемент> ::= <прописная буква> [<строчная буква>]
    # <прописная буква> ::= "A".."Z"
    # <строчная буква> ::= "a".."z"
    def chemElem():
        assert len(ch) == 1
        assert 'A' <= ch <= 'Z'

        elem = ch
        nextChar()

        if 'a' <= ch <= 'z':
            elem += ch
            nextChar()

        return elem

    # <часть> ::= <химический элемент> | "(" <последовательность> ")"
    def part():
        if ch == '(':
            nextChar()
            cnt = seq()

            assert ch == ')'
            nextChar()
            return cnt
        else:
            elem = chemElem()
            return Counter({ elem: 1 })

    def mul(cnt, num):
        for key in cnt:
            cnt[key] *= num

    # <последовательность> ::= <часть> <число> {<часть> <число>}
    def seq():
        cnt = part()
        num = number()
        mul(cnt, num)

        while ch == '(' or 'A' <= ch <= 'Z':
            cnt2 = part()
            num2 = number()
            mul(cnt2, num2)

            cnt += cnt2

        return cnt

    # <формула> ::= <число> <последовательность> {"+" <число> <последовательность>}
    def formula():
        num = number()
        cnt = seq()
        mul(cnt, num)

        while ch == '+':
            nextChar()
            num2 = number()
            cnt2 = seq()
            mul(cnt2, num2)

            cnt += cnt2

        return cnt

    nextChar()
    cnt = formula()

    assert ch == EOS

    return (s, cnt)

leftS, leftCnt = readFormula()
n = int(input())

for _ in range(n):
    rightS, rightCnt = readFormula()

    if leftCnt == rightCnt:
        print(leftS + '==' + rightS)
    else:
        print(leftS + '!=' + rightS)

File: 1337/1337.js
const { nextInt, nextLine } = require("../utils/Scanner");

const assert = (e) => {
    if (!e) {
        throw new Error("assertion failed");
    }
};

class Counter {
    constructor(dict = new Map()) {
        assert(dict instanceof Map)
        this.counter = new Map(dict);
    }

    get(key) {
        return this.counter.get(key) ?? 0;
    }

    mul(num) {
        const result = new Map();

        for (const [key, value] of this.counter) {
            result.set(key, value * num);
        }

        return new Counter(result);
    }

    add(other) {
        const result = new Map();

        for (const [key, value] of this.counter) {
            result.set(key, value);
        }

        for (const [key, value] of other.counter) {
            result.set(key, (result.get(key) ?? 0) + value);
        }

        return new Counter(result);
    }

    equals(other) {
        const keys = this.counter.keys().toArray();

        if (keys.length !== other.counter.keys().toArray) {
            return false;
        }

        for (const key of keys) {
            if (this.get(key) !== other.get(key)) {
                return false;
            }
        }

        return true;
    }
}

/*
<формула> ::= <число> <последовательность> {"+" <число> <последовательность>}
<последовательность> ::= <часть> <число> {<часть> <число>}
<часть> ::= <химический элемент> | "(" <последовательность> ")"
<химический элемент> ::= <прописная буква> [<строчная буква>]
<прописная буква> ::= "A".."Z"
<строчная буква> ::= "a".."z"
<число> ::= ["1".."9" {"0".."9"}]
*/

const readFormula = () => {
    const s = nextLine();
    let pos = 0;
    let ch = "";
    const EOS = "\n"; // end of string

    const nextChar = () => {
        if (pos < s.length) {
            ch = s[pos];
            pos++;
        } else {
            ch = EOS;
        }
    };

    // <число> ::= ["1".."9" {"0".."9"}]
    const number = () => {
        if (ch.length === 1 && "1" <= ch && ch <= "9") {
            let num = Number(ch);
            nextChar();

            while ("0" <= ch && ch <= "9") {
                num = num * 10 + Number(ch);
                assert(num <= 10_000);
                nextChar();
            }

            return num;
        } else {
            return 1;
        }
    };

    // <химический элемент> ::= <прописная буква> [<строчная буква>]
    // <прописная буква> ::= "A".."Z"
    // <строчная буква> ::= "a".."z"
    const chemElem = () => {
        assert(ch.length === 1 && "A" <= ch && ch <= "Z");

        let elem = ch;
        nextChar();

        if ("a" <= ch && ch <= "z") {
            elem += ch;
            nextChar();
        }

        return elem;
    };

    // <часть> ::= <химический элемент> | "(" <последовательность> ")"
    const part = () => {
        if (ch === "(") {
            nextChar();
            const cnt = seq();

            assert(ch === ")");
            nextChar();

            return cnt;
        } else {
            const elem = chemElem();

            return new Counter(new Map([[elem, 1]]));
        }
    };

    // <последовательность> ::= <часть> <число> {<часть> <число>}
    const seq = () => {
        let cnt = part();
        const num = number();
        cnt = cnt.mul(num);

        while (ch === "(" || ("A" <= ch && ch <= "Z")) {
            let cnt2 = part();
            const num2 = number();
            cnt2 = cnt2.mul(num2);

            cnt = cnt.add(cnt2);
        }

        return cnt;
    };

    // <формула> ::= <число> <последовательность> {"+" <число> <последовательность>}
    const formula = () => {
        const num = number();
        let s = seq();
        s = s.mul(num);

        while (ch === "+") {
            nextChar();
            const num1 = number();
            let s1 = seq();
            s1 = s1.mul(num1);

            s = s.add(s1);
        }

        return s;
    };

    nextChar();
    const cnt = formula();
    assert(ch === EOS);

    return { s, cnt };
};

const { s: leftS, cnt: leftCnt } = readFormula();
const n = nextInt();
nextLine();

for (let i = 0; i < n; i++) {
    const { s: rightS, cnt: rightCnt } = readFormula();
    const equal = leftCnt.equals(rightCnt);

    console.log(`${leftS}${equal ? '==' : '!='}${rightS}`);
}
